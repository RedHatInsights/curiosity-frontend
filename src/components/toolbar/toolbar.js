import React from 'react';
import PropTypes from 'prop-types';
import { Toolbar as PfToolbar, ToolbarContent, ToolbarFilter, ToolbarGroup } from '@patternfly/react-core';
import { Select } from '../form/select';
import { reduxTypes, store } from '../../redux';
import { rhsmApiTypes } from '../../types/rhsmApiTypes';
import { toolbarTypes } from './toolbarTypes';
import { helpers } from '../../common';
import { translate } from '../i18n/i18n';

/**
 * Application filter toolbar.
 *
 * @augments React.Component
 * @fires onClear
 * @fires onSlaSelect
 */
class Toolbar extends React.Component {
  /**
   * Clear filters' state.
   *
   * @event onClear
   */
  onClear = () => {
    this.dispatchFilter(reduxTypes.rhsm.SET_FILTER_SLA_RHSM, { [rhsmApiTypes.RHSM_API_QUERY_SLA]: null });
  };

  /**
   * Set SLA filter selection.
   *
   * @event onSlaSelect
   * @param {object} event
   */
  onSlaSelect = event => {
    const { value } = event;

    this.dispatchFilter(reduxTypes.rhsm.SET_FILTER_SLA_RHSM, { [rhsmApiTypes.RHSM_API_QUERY_SLA]: value });
  };

  /**
   * Dispatch a Redux store type.
   *
   * @param {string} type
   * @param {object} data
   */
  dispatchFilter(type, data = {}) {
    const { viewId } = this.props;

    if (type) {
      store.dispatch({
        type,
        viewId,
        ...data
      });
    }
  }

  // ToDo: API to provide SLA options from endpoint.
  /**
   * Available and selected SLA options.
   *
   * @returns {{slaOptionsSelected: Array, slaOptions: object}}
   */
  filterSla() {
    const { query } = this.props;

    const slaOptions = toolbarTypes.getOptions();
    const filterSla =
      typeof query[rhsmApiTypes.RHSM_API_QUERY_SLA] === 'string' &&
      slaOptions.options.find(val => val.value === query[rhsmApiTypes.RHSM_API_QUERY_SLA]);

    const slaOptionsSelected =
      (filterSla && filterSla.title && [filterSla.title]) || (slaOptions.selected && [slaOptions.selected]) || [];

    return { slaOptions, slaOptionsSelected };
  }

  /**
   * Render a filter toolbar.
   *
   * @returns {Node}
   */
  render() {
    const { isDisabled, t } = this.props;

    if (isDisabled) {
      return null;
    }

    const { slaOptions, slaOptionsSelected } = this.filterSla();

    return (
      <PfToolbar
        id="curiosity-toolbar"
        className="pf-m-toggle-group-container ins-c-primary-toolbar"
        collapseListedFiltersBreakpoint="xl"
        clearAllFilters={this.onClear}
      >
        <ToolbarContent>
          <ToolbarGroup variant="filter-group">
            <ToolbarFilter
              chips={slaOptionsSelected}
              deleteChip={this.onClear}
              categoryName={t('curiosity-toolbar.slaCategory')}
            >
              <Select
                aria-label={t('curiosity-toolbar.slaCategory')}
                onSelect={this.onSlaSelect}
                selectedOptions={slaOptionsSelected}
                placeholder={t('curiosity-toolbar.slaPlaceholder')}
                options={slaOptions.options}
              />
            </ToolbarFilter>
          </ToolbarGroup>
        </ToolbarContent>
      </PfToolbar>
    );
  }
}

/**
 * Prop types
 *
 * @type {{viewId: string, t: Function, query: object, isDisabled: boolean }}
 */
Toolbar.propTypes = {
  query: PropTypes.shape({
    [rhsmApiTypes.RHSM_API_QUERY_SLA]: PropTypes.string
  }),
  isDisabled: PropTypes.bool,
  t: PropTypes.func,
  viewId: PropTypes.string
};

/**
 * Default props.
 *
 * @type {{viewId: string, t: translate, query: {}, isDisabled: boolean}}
 */
Toolbar.defaultProps = {
  query: {},
  isDisabled: helpers.UI_DISABLED_TOOLBAR,
  t: translate,
  viewId: 'toolbar'
};

export { Toolbar as default, Toolbar };
