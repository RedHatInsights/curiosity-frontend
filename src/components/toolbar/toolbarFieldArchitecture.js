import React from 'react';
import PropTypes from 'prop-types';
import { useProduct, useProductToolbarQuery } from '../productView/productViewContext';
import { reduxTypes, storeHooks } from '../../redux';
import { RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';
import { Select, SelectPosition } from '../form/select';
import { translate } from '../i18n/i18n';

/**
 * Generate select field options from config
 *
 * @param {object} options
 * @param {Function} options.useProduct
 * @returns {Function}
 */
const useToolbarFieldOptions = ({ useProduct: useAliasProduct = useProduct } = {}) => {
  const { productArchitectures } = useAliasProduct();
  const options = [];

  if (Array.isArray(productArchitectures)) {
    options.push(
      ...productArchitectures.map(type => ({
        title: translate('curiosity-toolbar.label', {
          context: ['architecture', (type === '' && 'unspecified') || type]
        }),
        value: type,
        selected: false
      }))
    );
  }

  return options;
};

/**
 * On select update.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useProduct
 * @returns {Function}
 */
const useOnSelect = ({
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useProduct: useAliasProduct = useProduct
} = {}) => {
  const { viewId } = useAliasProduct();
  const dispatch = useAliasDispatch();

  return ({ value = null } = {}) => {
    dispatch([
      {
        type: reduxTypes.query.SET_QUERY_RESET_INVENTORY_LIST,
        viewId
      },
      {
        type: reduxTypes.query.SET_QUERY,
        viewId,
        filter: RHSM_API_QUERY_SET_TYPES.ARCHITECTURE,
        value
      }
    ]);
  };
};

/**
 * Display an architecture field with generated options.
 *
 * @fires onSelect
 * @param {object} props
 * @param {boolean} props.isFilter
 * @param {string} props.position
 * @param {Function} props.t
 * @param {Function} props.useOnSelect
 * @param {Function} props.useProductToolbarQuery
 * @param {Function} props.useToolbarFieldOptions
 * @returns {Node}
 */
const ToolbarFieldArchitecture = ({
  isFilter,
  position,
  t,
  useOnSelect: useAliasOnSelect,
  useProductToolbarQuery: useAliasProductToolbarQuery,
  useToolbarFieldOptions: useAliasToolbarFieldOptions
}) => {
  const { [RHSM_API_QUERY_SET_TYPES.ARCHITECTURE]: updatedValue } = useAliasProductToolbarQuery();
  const onSelect = useAliasOnSelect();
  const options = useAliasToolbarFieldOptions();
  const updatedOptions = options.map(option => ({ ...option, selected: option.value === updatedValue }));

  return (
    <Select
      aria-label={t('curiosity-toolbar.placeholder', { context: [isFilter && 'filter', 'architecture'] })}
      onSelect={onSelect}
      options={updatedOptions}
      selectedOptions={updatedValue}
      placeholder={t('curiosity-toolbar.placeholder', { context: [isFilter && 'filter', 'architecture'] })}
      position={position}
      data-test="toolbarFieldArchitecture"
    />
  );
};

/**
 * Prop types.
 *
 * @type {{useOnSelect: Function, t: Function, useProductToolbarQuery: Function, isFilter: boolean,
 *     position: string, useToolbarFieldOptions: Function}}
 */
ToolbarFieldArchitecture.propTypes = {
  isFilter: PropTypes.bool,
  position: PropTypes.string,
  t: PropTypes.func,
  useOnSelect: PropTypes.func,
  useProductToolbarQuery: PropTypes.func,
  useToolbarFieldOptions: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{useOnSelect: Function, t: translate, useProductToolbarQuery: Function, isFilter: boolean,
 *     position: string, useToolbarFieldOptions: Function}}
 */
ToolbarFieldArchitecture.defaultProps = {
  isFilter: false,
  position: SelectPosition.left,
  t: translate,
  useOnSelect,
  useProductToolbarQuery,
  useToolbarFieldOptions
};

export { ToolbarFieldArchitecture as default, ToolbarFieldArchitecture, useOnSelect, useToolbarFieldOptions };
