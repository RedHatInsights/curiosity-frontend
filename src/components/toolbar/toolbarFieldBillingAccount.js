import React, { useEffect, useMemo } from 'react';
import { reduxTypes, storeHooks } from '../../redux';
import { useProduct, useProductQuery } from '../productView/productViewContext';
import { Select, SelectPosition } from '../form/select';
import { RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';
import { translate } from '../i18n/i18n';

/**
 * A dynamic Billing Account select filter.
 *
 * @memberof Toolbar
 * @module ToolbarFieldBillingAccount
 */

/**
 * ToDo: review activating the default account to help in auto-population.
 */
/**
 * Generate select field options from service.
 * Note: Avoid using the translation wrapper for "account" for scenarios where the account
 * contains underscores. Underscores interfere with the contextual string lookup.
 *
 * @param {object} options
 * @param {useProduct} [options.useProduct=useProduct]
 * @param {useProductQuery} [options.useProductQuery=useProductQuery]
 * @param {storeHooks.reactRedux.useSelector} [options.useSelector=storeHooks.reactRedux.useSelector]
 * @returns {Array<{title: React.ReactNode, value: string, isSelected: boolean}>}
 */
const useToolbarFieldOptions = ({
  useProduct: useAliasProduct = useProduct,
  useProductQuery: useAliasProductQuery = useProductQuery,
  useSelector: useAliasSelector = storeHooks.reactRedux.useSelector
} = {}) => {
  const { [RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER]: billingProvider } = useAliasProductQuery();
  const { productId } = useAliasProduct();
  const { data = {} } = useAliasSelector(({ app }) => app.billingAccounts?.[productId], {});
  // const updatedBillingProvider = billingProvider || data?.defaultProvider;
  const updatedBillingProvider = billingProvider;
  const defaultAccount = data?.defaultAccountByProvider?.[updatedBillingProvider];
  const billingAccounts = data?.accountsByProvider?.[updatedBillingProvider];

  return useMemo(
    () =>
      billingAccounts?.map(account => ({
        title: account,
        value: account,
        isSelected: account === defaultAccount
      })) || [],
    [billingAccounts, defaultAccount]
  );
};

/**
 * On select update billing account.
 *
 * @param {object} options
 * @param {storeHooks.reactRedux.useDispatch} [options.useDispatch=storeHooks.reactRedux.useDispatch]
 * @param {useProduct} [options.useProduct=useProduct]
 * @returns {Function}
 */
const useOnSelect = ({
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useProduct: useAliasProduct = useProduct
} = {}) => {
  const { productId } = useAliasProduct();
  const dispatch = useAliasDispatch();

  return ({ value = null } = {}) => {
    dispatch([
      {
        type: reduxTypes.query.SET_QUERY_RESET_INVENTORY_LIST,
        viewId: productId
      },
      {
        type: reduxTypes.query.SET_QUERY,
        viewId: productId,
        filter: RHSM_API_QUERY_SET_TYPES.BILLING_ACCOUNT_ID,
        value
      }
    ]);
  };
};

/**
 * Display a billing provider field with options.
 *
 * @param {object} props
 * @param {boolean} [props.isFilter=false]
 * @param {SelectPosition} [props.position=SelectPosition.left]
 * @param {translate} [props.t=translate]
 * @param {useOnSelect} [props.useOnSelect=useOnSelect]
 * @param {useProductQuery} [props.useProductQuery=useProductQuery]
 * @param {useToolbarFieldOptions} [props.useToolbarFieldOptions=useToolbarFieldOptions]
 * @fires onSelect
 * @returns {JSX.Element}
 */
const ToolbarFieldBillingAccount = ({
  isFilter = false,
  position = SelectPosition.left,
  t = translate,
  useOnSelect: useAliasOnSelect = useOnSelect,
  useProductQuery: useAliasProductQuery = useProductQuery,
  useToolbarFieldOptions: useAliasToolbarFieldOptions = useToolbarFieldOptions
}) => {
  const {
    [RHSM_API_QUERY_SET_TYPES.BILLING_ACCOUNT_ID]: updatedValue,
    [RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER]: billingProvider
  } = useAliasProductQuery();
  const onSelect = useAliasOnSelect();
  const options = useAliasToolbarFieldOptions();
  const updatedOptions = options.map(option => ({
    ...option,
    isSelected: option.value === updatedValue
  }));

  // Select an account on provider update
  useEffect(() => {
    const selectedAccount = updatedOptions.find(({ isSelected }) => isSelected === true)?.value;

    if (selectedAccount) {
      onSelect({ value: selectedAccount });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [billingProvider]);

  return (
    <Select
      isReadOnly={updatedOptions.length === 1}
      aria-label={t(`curiosity-toolbar.placeholder${(isFilter && '_filter') || ''}`, { context: 'billing_account' })}
      onSelect={onSelect}
      options={updatedOptions}
      selectedOptions={updatedValue}
      placeholder={t(`curiosity-toolbar.placeholder${(isFilter && '_filter') || ''}`, { context: 'billing_account' })}
      alignment={{ position }}
      data-test="toolbarFieldBillingAccount"
    />
  );
};

export { ToolbarFieldBillingAccount as default, ToolbarFieldBillingAccount, useOnSelect, useToolbarFieldOptions };
