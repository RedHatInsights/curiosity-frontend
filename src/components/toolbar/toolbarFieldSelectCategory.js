import React from 'react';
import PropTypes from 'prop-types';
import { FilterIcon } from '@patternfly/react-icons';
import { useShallowCompareEffect } from 'react-use';
import { reduxTypes, storeHooks } from '../../redux';
import { useProduct, useProductToolbarConfig } from '../productView/productViewContext';
import { Select } from '../form/select';
import { RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';
import { translate } from '../i18n/i18n';
import { ToolbarFieldGranularity, toolbarFieldOptions as granularityOptions } from './toolbarFieldGranularity';
import { ToolbarFieldRangedMonthly, toolbarFieldOptions as rangedMonthlyOptions } from './toolbarFieldRangedMonthly';
import {
  ToolbarFieldBillingProvider,
  toolbarFieldOptions as billingProviderOptions
} from './toolbarFieldBillingProvider';
import { ToolbarFieldSla, toolbarFieldOptions as slaOptions } from './toolbarFieldSla';
import { ToolbarFieldUom, toolbarFieldOptions as uomOptions } from './toolbarFieldUom';
import { ToolbarFieldUsage, toolbarFieldOptions as usageOptions } from './toolbarFieldUsage';

/**
 * Select field options. Use function instead of arrow func to help with component
 * display name during testing.
 *
 * @type {{title: (string|Node), value: string, selected: boolean}[]}
 */
const toolbarFieldOptions = [
  {
    title: translate('curiosity-toolbar.category', { context: RHSM_API_QUERY_SET_TYPES.GRANULARITY }),
    value: RHSM_API_QUERY_SET_TYPES.GRANULARITY,
    component: function Granularity(props) {
      return <ToolbarFieldGranularity key="selectCategory_granularity" {...props} />;
    },
    options: granularityOptions,
    isClearable: false
  },
  {
    title: translate('curiosity-toolbar.category', { context: 'rangedMonthly' }),
    value: 'rangedMonthly',
    component: function RangedMonthly(props) {
      return <ToolbarFieldRangedMonthly key="selectCategory_rangedMonthly" {...props} />;
    },
    options: rangedMonthlyOptions,
    isClearable: false
  },
  {
    title: translate('curiosity-toolbar.category', { context: RHSM_API_QUERY_SET_TYPES.UOM }),
    value: RHSM_API_QUERY_SET_TYPES.UOM,
    component: function Uom(props) {
      return <ToolbarFieldUom key="selectCategory_uom" {...props} />;
    },
    options: uomOptions,
    isClearable: false
  },
  {
    title: translate('curiosity-toolbar.category', { context: RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER }),
    value: RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER,
    component: function BillingProvider(props) {
      return <ToolbarFieldBillingProvider key="selectCategory_billingProvider" {...props} />;
    },
    options: billingProviderOptions,
    isClearable: true
  },
  {
    title: translate('curiosity-toolbar.category', { context: RHSM_API_QUERY_SET_TYPES.SLA }),
    value: RHSM_API_QUERY_SET_TYPES.SLA,
    component: function Sla(props) {
      return <ToolbarFieldSla key="selectCategory_sla" {...props} />;
    },
    options: slaOptions,
    isClearable: true
  },
  {
    title: translate('curiosity-toolbar.category', { context: RHSM_API_QUERY_SET_TYPES.USAGE }),
    value: RHSM_API_QUERY_SET_TYPES.USAGE,
    component: function Usage(props) {
      return <ToolbarFieldUsage key="selectCategory_usage" {...props} />;
    },
    options: usageOptions,
    isClearable: true
  }
].map(option => ({
  ...option,
  selected: false
}));

/**
 * On select update category.
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @param {Function} options.useProduct
 * @returns {Function}
 */
const useOnSelect = ({
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useProduct: useAliasProduct = useProduct
} = {}) => {
  const { viewId } = useAliasProduct();
  const dispatch = useAliasDispatch();

  return ({ value = null } = {}) => {
    dispatch([
      {
        type: reduxTypes.toolbar.SET_FILTER_TYPE,
        viewId,
        currentFilter: value
      }
    ]);
  };
};

/**
 * Return filtered category options, current, and initial value.
 *
 * @param {object} options
 * @param {Array} options.categoryOptions
 * @param {Function} options.useProduct
 * @param {Function} options.useProductToolbarConfig
 * @param {Function} options.useSelector
 * @returns {object}
 */
const useSelectCategoryOptions = ({
  categoryOptions = toolbarFieldOptions,
  useProduct: useAliasProduct = useProduct,
  useProductToolbarConfig: useAliasProductToolbarConfig = useProductToolbarConfig,
  useSelector: useAliasSelector = storeHooks.reactRedux.useSelector
} = {}) => {
  const { viewId } = useAliasProduct();
  const { currentFilter: updatedValue } = useAliasSelector(({ toolbar }) => toolbar.filters?.[viewId], {});
  const { filters = [] } = useAliasProductToolbarConfig();

  let initialValue;

  const updatedOptions = filters.map(({ id, selected }) => {
    const option = categoryOptions.find(({ value }) => id === value);

    if (updatedValue === undefined && selected) {
      initialValue = option.value;
    }

    return {
      ...option,
      selected: (updatedValue === undefined && selected) || updatedValue === option.value
    };
  });

  return {
    currentCategory: updatedValue,
    initialCategory: initialValue,
    options: updatedOptions
  };
};

/**
 * Display a granularity field with options.
 *
 * @fires onSelect
 * @param {object} props
 * @param {Function} props.t
 * @param {Function} props.useOnSelect
 * @param {Function} props.useSelectCategoryOptions
 * @returns {Node}
 */
const ToolbarFieldSelectCategory = ({
  t,
  useOnSelect: useAliasOnSelect,
  useSelectCategoryOptions: useAliasSelectCategoryOptions
}) => {
  const { currentCategory: updatedValue, initialCategory: initialValue, options } = useAliasSelectCategoryOptions();
  const onSelect = useAliasOnSelect();

  useShallowCompareEffect(() => {
    if (initialValue) {
      onSelect({ value: initialValue });
    }
  }, [initialValue, onSelect]);

  return (
    <Select
      aria-label={t('curiosity-toolbar.placeholder_filter')}
      onSelect={onSelect}
      options={options}
      selectedOptions={updatedValue}
      placeholder={t('curiosity-toolbar.placeholder_filter')}
      toggleIcon={<FilterIcon />}
      data-test="toolbarFieldCategory"
    />
  );
};

/**
 * Prop types.
 *
 * @type {{useOnSelect: Function, t: Function, useSelectCategoryOptions: Function}}
 */
ToolbarFieldSelectCategory.propTypes = {
  t: PropTypes.func,
  useOnSelect: PropTypes.func,
  useSelectCategoryOptions: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{useOnSelect: Function, t: Function, useSelectCategoryOptions: Function}}
 */
ToolbarFieldSelectCategory.defaultProps = {
  t: translate,
  useOnSelect,
  useSelectCategoryOptions
};

export {
  ToolbarFieldSelectCategory as default,
  ToolbarFieldSelectCategory,
  toolbarFieldOptions,
  useOnSelect,
  useSelectCategoryOptions
};
