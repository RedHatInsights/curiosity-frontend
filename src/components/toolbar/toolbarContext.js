import React from 'react';
import { ToolbarItem } from '@patternfly/react-core';
import { useProductQuery, useProductGraphTallyQuery, useProductToolbarConfig } from '../productView/productViewContext';
import { RHSM_API_QUERY_SET_TYPES } from '../../services/rhsm/rhsmConstants';
import { useOnSelect as useSelectCategoryOnSelect, toolbarFieldOptions } from './toolbarFieldSelectCategory';
import { useOnSelect as useBillingProviderOnSelect } from './toolbarFieldBillingProvider';
import { useOnSelect as useSlaOnSelect } from './toolbarFieldSla';
import { useOnSelect as useUsageOnSelect } from './toolbarFieldUsage';
import { SelectPosition } from '../form/select';
import { helpers } from '../../common/helpers';

/**
 * Return current values for categories/queries.
 *
 * @param {object} options
 * @param {Function} options.useProductQuery
 * @param {Function} options.useProductGraphTallyQuery
 * @returns {object}
 */
const useToolbarFieldQueries = ({
  useProductQuery: useAliasProductQuery = useProductQuery,
  useProductGraphTallyQuery: useAliasProductGraphTallyQuery = useProductGraphTallyQuery
} = {}) => {
  const {
    [RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER]: billingProvider,
    [RHSM_API_QUERY_SET_TYPES.SLA]: sla,
    [RHSM_API_QUERY_SET_TYPES.UOM]: uom,
    [RHSM_API_QUERY_SET_TYPES.USAGE]: usage
  } = useAliasProductQuery();
  const { [RHSM_API_QUERY_SET_TYPES.GRANULARITY]: granularity } = useAliasProductGraphTallyQuery();

  return {
    [RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER]: billingProvider,
    [RHSM_API_QUERY_SET_TYPES.GRANULARITY]: granularity,
    [RHSM_API_QUERY_SET_TYPES.SLA]: sla,
    [RHSM_API_QUERY_SET_TYPES.UOM]: uom,
    [RHSM_API_QUERY_SET_TYPES.USAGE]: usage
  };
};

/**
 * Clear a specific toolbar category.
 *
 * @param {object} options
 * @param {Function} options.useBillingProviderOnSelect
 * @param {Function} options.useSlaOnSelect
 * @param {Function} options.useUsageOnSelect
 * @returns {Function}
 */
const useToolbarFieldClear = ({
  useBillingProviderOnSelect: useAliasBillingProviderOnSelect = useBillingProviderOnSelect,
  useSlaOnSelect: useAliasSlaOnSelect = useSlaOnSelect,
  useUsageOnSelect: useAliasUsageOnSelect = useUsageOnSelect
} = {}) => {
  const billingOnSelect = useAliasBillingProviderOnSelect();
  const slaOnSelect = useAliasSlaOnSelect();
  const usageOnSelect = useAliasUsageOnSelect();

  return field => {
    switch (field) {
      case RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER:
        billingOnSelect();
        break;
      case RHSM_API_QUERY_SET_TYPES.SLA:
        slaOnSelect();
        break;
      case RHSM_API_QUERY_SET_TYPES.USAGE:
        usageOnSelect();
        break;
      default:
        break;
    }
  };
};

/**
 * Clear all available toolbar categories.
 *
 * @param {object} options
 * @param {Function} options.useProductQuery
 * @param {Function} options.useSelectCategoryOnSelect
 * @param {Function} options.useBillingProviderOnSelect
 * @param {Function} options.useSlaOnSelect
 * @param {Function} options.useUsageOnSelect
 * @returns {Function}
 */
const useToolbarFieldClearAll = ({
  useProductQuery: useAliasProductQuery = useProductQuery,
  useSelectCategoryOnSelect: useAliasSelectCategoryOnSelect = useSelectCategoryOnSelect,
  useBillingProviderOnSelect: useAliasBillingProviderOnSelect = useBillingProviderOnSelect,
  useSlaOnSelect: useAliasSlaOnSelect = useSlaOnSelect,
  useUsageOnSelect: useAliasUsageOnSelect = useUsageOnSelect
} = {}) => {
  const {
    [RHSM_API_QUERY_SET_TYPES.BILLING_PROVIDER]: billingProvider,
    [RHSM_API_QUERY_SET_TYPES.SLA]: sla,
    [RHSM_API_QUERY_SET_TYPES.USAGE]: usage
  } = useAliasProductQuery();
  const billingOnSelect = useAliasBillingProviderOnSelect();
  const slaOnSelect = useAliasSlaOnSelect();
  const usageOnSelect = useAliasUsageOnSelect();
  const selectCategoryOnSelect = useAliasSelectCategoryOnSelect();

  return hardFilterReset => {
    if (typeof billingProvider === 'string') {
      billingOnSelect();
    }

    if (typeof sla === 'string') {
      slaOnSelect();
    }

    if (typeof usage === 'string') {
      usageOnSelect();
    }

    if (hardFilterReset) {
      selectCategoryOnSelect();
    }
  };
};

/**
 * Return list of secondary toolbar fields for display.
 *
 * @param {object} options
 * @param {Array} options.categoryOptions
 * @param {Function} options.useProductToolbarConfig
 * @returns {Array}
 */
const useToolbarSecondaryFields = ({
  categoryOptions = toolbarFieldOptions,
  useProductToolbarConfig: useAliasProductToolbarConfig = useProductToolbarConfig
} = {}) => {
  const { secondaryFilters = [] } = useAliasProductToolbarConfig();

  return secondaryFilters.map(({ id, content }) => {
    const option = categoryOptions.find(({ value: categoryOptionValue }) => id === categoryOptionValue);
    const { component: OptionComponent } = option || {};

    return (
      (OptionComponent && (
        <ToolbarItem key={`option-${id}`}>
          <OptionComponent isFilter={false} position={SelectPosition.right} />
        </ToolbarItem>
      )) || <ToolbarItem key={helpers.generateId()}>{content}</ToolbarItem> ||
      null
    );
  });
};

const context = {
  useToolbarFieldClear,
  useToolbarFieldClearAll,
  useToolbarFieldQueries,
  useToolbarSecondaryFields
};

export {
  context as default,
  context,
  useToolbarFieldClear,
  useToolbarFieldClearAll,
  useToolbarFieldQueries,
  useToolbarSecondaryFields
};
