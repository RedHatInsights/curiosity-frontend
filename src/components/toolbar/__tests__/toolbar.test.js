import React from 'react';
import { ToolbarFilter } from '@patternfly/react-core';
import { Toolbar } from '../toolbar';
import { RHSM_API_QUERY_SLA_TYPES, RHSM_API_QUERY_SET_TYPES } from '../../../services/rhsm/rhsmConstants';
import { toolbarFieldOptions as selectCategoryOptions } from '../toolbarFieldSelectCategory';

describe('Toolbar Component', () => {
  it('should render a basic component', async () => {
    const props = {
      useSelectCategoryOptions: () => ({ options: [selectCategoryOptions[4], selectCategoryOptions[5]] })
    };
    const component = await shallowHookComponent(<Toolbar {...props} />);

    expect(component).toMatchSnapshot('basic');
  });

  it('should return an empty render when disabled or missing filters', async () => {
    const props = {
      isDisabled: true
    };
    const component = await shallowHookComponent(<Toolbar {...props} />);

    expect(component).toMatchSnapshot('disabled component');

    component.setProps({
      isDisabled: false,
      useSelectCategoryOptions: () => ({ options: [] }),
      useToolbarFields: () => ({ itemFields: [], secondaryFields: [] })
    });
    expect(component).toMatchSnapshot('missing filters');

    component.setProps({
      useToolbarFields: () => ({ itemFields: [], secondaryFields: [<span key="lorem">lorem ipsum</span>] })
    });
    expect(component).toMatchSnapshot('missing primary, has secondary filters');
  });

  it('should hide categories when a single filter is available', async () => {
    const props = {
      useSelectCategoryOptions: () => ({ options: [selectCategoryOptions[4]] })
    };
    const component = await shallowHookComponent(<Toolbar {...props} />);

    expect(component).toMatchSnapshot('single filter');
  });

  it('should handle updating toolbar chips', async () => {
    const props = {
      useSelectCategoryOptions: () => ({ options: [selectCategoryOptions[4]] }),
      useProductToolbarQuery: () => ({ [RHSM_API_QUERY_SET_TYPES.SLA]: RHSM_API_QUERY_SLA_TYPES.PREMIUM })
    };
    const component = await shallowHookComponent(<Toolbar {...props} />);

    expect(component.find(ToolbarFilter).props()).toMatchSnapshot('chips');

    component.setProps({
      useSelectCategoryOptions: () => ({ options: [{ ...selectCategoryOptions[4], isClearable: false }] })
    });

    expect(component.find(ToolbarFilter).props()).toMatchSnapshot('chips, not clearable');
  });

  it('should handle displaying secondary components, fields', async () => {
    const props = {
      useSelectCategoryOptions: () => ({ options: [selectCategoryOptions[4]] }),
      useToolbarFields: () => ({ itemFields: [], secondaryFields: [<span key="lorem">lorem ipsum</span>] })
    };
    const component = await shallowHookComponent(<Toolbar {...props} />);

    expect(component).toMatchSnapshot('secondary');
  });

  it('should handle displaying a group variant field', async () => {
    const props = {
      isGroupVariantDisabled: false,
      useSelectCategoryOptions: () => ({ options: [selectCategoryOptions[4]] }),
      useToolbarFields: () => ({ itemFields: [], secondaryFields: [<span key="lorem">lorem ipsum</span>] })
    };
    const component = await shallowHookComponent(<Toolbar {...props} />);

    expect(component).toMatchSnapshot('group variant');

    component.setProps({
      useSelectCategoryOptions: () => ({ options: [] }),
      useToolbarFields: () => ({ itemFields: [], secondaryFields: [] })
    });

    expect(component).toMatchSnapshot('group variant, standalone');
  });
});
