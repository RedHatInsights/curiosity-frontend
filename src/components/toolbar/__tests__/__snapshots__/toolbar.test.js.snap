// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar Component should handle updating state and dispatching an sla filter: clear filters 1`] = `null`;

exports[`Toolbar Component should handle updating state and dispatching an sla filter: sla selected 1`] = `null`;

exports[`Toolbar Component should render a non-connected component: non-connected 1`] = `
<Toolbar
  className="pf-m-toggle-group-container ins-c-primary-toolbar"
  clearAllFilters={[Function]}
  collapseListedFiltersBreakpoint="xl"
  id="curiosity-toolbar"
>
  <ToolbarContent
    isExpanded={false}
    showClearFiltersButton={false}
  >
    <ForwardRef
      variant="filter-group"
    >
      <ToolbarFilter
        categoryName="t(curiosity-toolbar.slaCategory)"
        chips={Array []}
        deleteChip={[Function]}
        showToolbarItem={true}
      >
        <Select
          aria-label="t(curiosity-toolbar.slaCategory)"
          ariaLabel="Select option"
          className=""
          id="generatedid-"
          isDisabled={false}
          name={null}
          onSelect={[Function]}
          options={
            Array [
              Object {
                "title": "t(curiosity-toolbar.slaPremium)",
                "value": "premium",
              },
              Object {
                "title": "t(curiosity-toolbar.slaStandard)",
                "value": "standard",
              },
              Object {
                "title": "t(curiosity-toolbar.slaSelfSupport)",
                "value": "self-support",
              },
              Object {
                "title": "t(curiosity-toolbar.slaNone)",
                "value": "",
              },
            ]
          }
          placeholder="t(curiosity-toolbar.slaPlaceholder)"
          selectedOptions={Array []}
          variant="single"
        />
      </ToolbarFilter>
    </ForwardRef>
  </ToolbarContent>
</Toolbar>
`;

exports[`Toolbar Component should return an empty render when disabled: disabled component 1`] = `""`;
